test/unit/ReservoirRouter.t.sol

```diff
@@ -0,0 +1,759 @@
+            _router.refundETH,
+            ()
+        ));
+        bytes[] memory lResult = _router.multicall{value: lAmounts[0] }(_data);
+
+        // assert
+        uint256[] memory lAmountsReturned = abi.decode(lResult[0], (uint256[]));
+        assertEq(_tokenB.balanceOf(address(this)), lAmtOut);
+        assertEq(lAmountsReturned, lAmounts);
+    }
+
+    function testSwapVariableForExact_NativeIn_RefundETH() public
+    {
+        // arrange
+        testAddLiquidity_CreatePair_CP_Native();
+        delete _data;
+        vm.stopPrank();
+        uint256 lAmtOut = 1e18;
+        uint256 lEtherToMint = 100 ether;
+        deal(address(this), lEtherToMint);
+
+        // ETH -> tokenA -> tokenB
+        address[] memory lPath = new address[](3);
+        lPath[0] = address(_weth);
+        lPath[1] = address(_tokenA);
+        lPath[2] = address(_tokenB);
+        uint256[] memory lCurveIds = new uint256[](2);
+        lCurveIds[0] = 0;
+        lCurveIds[1] = 0;
+
+        uint256[] memory lAmounts = _router.getAmountsIn(lAmtOut, lPath, lCurveIds);
+        uint256 lAmountInMax = lAmounts[0] * 101 / 100; // 1% slippage
+
+        _data.push(abi.encodeCall(
+            _router.swapVariableForExact,
+            (lAmtOut, lAmountInMax, lPath, lCurveIds, address(this))
+        ));
+        _data.push(abi.encodeCall(
+            _router.refundETH,
+            ()
+        ));
+
+        // send way too much ETH to the router
+        bytes[] memory lResult = _router.multicall{ value: address(this).balance }(_data);
+
+        // assert
+        assertEq(address(this).balance, lEtherToMint - lAmounts[0]);
+        assertEq(address(_router).balance, 0);
+    }
+
+    function testSwapVariableForExact_NativeOut() public
+    {
+        testAddLiquidity_CreatePair_CP_Native();
+        delete _data;
+        vm.stopPrank();
+        deal(address(this), 0);
+        uint256 lAmtOut = 1e6;
+
+        // tokenB -> tokenA -> ETH
+        address[] memory lPath = new address[](3);
+        lPath[0] = address(_tokenB);
+        lPath[1] = address(_tokenA);
+        lPath[2] = address(_weth);
+        uint256[] memory lCurveIds = new uint256[](2);
+        lCurveIds[0] = 0;
+        lCurveIds[1] = 0;
+        uint256[] memory lAmounts = _router.getAmountsIn(lAmtOut, lPath, lCurveIds);
+        uint256 lAmountInMax = lAmounts[0] * 101 / 100; // 1% slippage
+        _tokenB.mint(address(this), lAmounts[0]);
+        _tokenB.approve(address(_router), lAmounts[0]);
+
+        // act
+        _data.push(abi.encodeCall(
+            _router.swapVariableForExact,
+            (lAmtOut, lAmountInMax, lPath, lCurveIds, address(_router))
+        ));
+        _data.push(abi.encodeCall(
+            _router.unwrapWETH,
+            (lAmounts[lAmounts.length - 1], address(this))
+        ));
+
+        bytes[] memory lResult = _router.multicall(_data);
+
+        // assert
+        assertEq(address(this).balance, lAmtOut);
+    }
+}
```

Comment:
Some additional tests:

- testSwapMixedCurves

Reply:
added as a todo in the test files, for both `swapExact` and `swapVariable`